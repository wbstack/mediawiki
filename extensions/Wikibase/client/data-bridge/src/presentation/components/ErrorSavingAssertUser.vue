<template>
	<div class="wb-db-error-saving-assertuser">
		<h2
			class="wb-db-error-saving-assertuser__heading"
		>
			{{ $messages.getText( $messages.KEYS.SAVING_ERROR_ASSERTUSER_HEADING ) }}
		</h2>
		<IconMessageBox
			class="wb-db-error-saving-assertuser__message"
			type="warning"
		>
			{{ $messages.getText( $messages.KEYS.SAVING_ERROR_ASSERTUSER_MESSAGE ) }}
		</IconMessageBox>
		<div class="wb-db-error-saving-assertuser__buttons">
			<EventEmittingButton
				class="wb-db-error-saving-assertuser__proceed"
				type="primaryProgressive"
				size="M"
				:message="$messages.getText( publishOrSave )"
				@click="proceed"
			/>
			<EventEmittingButton
				class="wb-db-error-saving-assertuser__login"
				type="neutral"
				size="M"
				:href="loginUrl"
				:message="$messages.getText( $messages.KEYS.SAVING_ERROR_ASSERTUSER_LOGIN )"
				:new-tab="true"
				:prevent-default="false"
				@click="back"
			/>
			<EventEmittingButton
				class="wb-db-error-saving-assertuser__back"
				type="link"
				size="M"
				:message="$messages.getText( $messages.KEYS.SAVING_ERROR_ASSERTUSER_KEEP_EDITING )"
				@click="back"
			/>
		</div>
	</div>
</template>

<script lang="ts">
import { mixins } from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import StateMixin from '@/presentation/StateMixin';
import Component from 'vue-class-component';
import EventEmittingButton from '@/presentation/components/EventEmittingButton.vue';
import IconMessageBox from '@/presentation/components/IconMessageBox.vue';
import ReportIssue from '@/presentation/components/ReportIssue.vue';

/**
 * A component which gets shown when an error occurs while saving and the user is logged out.
 */
@Component( {
	components: {
		EventEmittingButton,
		IconMessageBox,
		ReportIssue,
	},
} )
export default class ErrorSavingAssertUser extends mixins( StateMixin ) {
	@Prop( { required: true, type: String } )
	public loginUrl!: string;

	public async proceed(): Promise<void> {
		await this.rootModule.dispatch( 'stopAssertingUserWhenSaving' );
		await this.rootModule.dispatch( 'retrySave' );
	}

	public back(): void {
		this.rootModule.dispatch( 'goBackFromErrorToReady' );
	}

	public get publishOrSave(): string {
		return this.$bridgeConfig.usePublish ?
			this.$messages.KEYS.SAVING_ERROR_ASSERTUSER_PUBLISH : this.$messages.KEYS.SAVING_ERROR_ASSERTUSER_SAVE;
	}

}
</script>

<style lang="scss">
.wb-db-error-saving-assertuser {
	display: flex;
	align-items: stretch;
	justify-content: center;
	flex-direction: column;

	&__heading {
		@include h3();
		@include marginForCenterColumnHeading();
		text-align: center;
	}

	&__message {
		@include marginForCenterColumn( 3 * $base-spacing-unit );
	}

	&__buttons {
		@include marginForCenterColumn( $margin-bottom: 3 * $base-spacing-unit );
		padding: 0 3 * $base-spacing-unit-fixed;

		> * {
			width: 100%;
		}
	}

	&__proceed,
	&__login {
		margin-bottom: $base-spacing-unit;
	}

	@media ( max-width: $breakpoint ) {
		&__buttons {
			margin-bottom: $base-spacing-unit;
			padding: 0;
		}

		&__proceed {
			margin-bottom: 2 * $base-spacing-unit;
		}
	}

	&__back {
		margin-right: $inter-button-spacing;

		@media ( max-width: $breakpoint ) {
			// show button in AppHeader instead
			display: none;
		}
	}
}
</style>
