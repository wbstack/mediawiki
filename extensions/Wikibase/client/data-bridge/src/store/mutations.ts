import { DataValue } from '@wmde/wikibase-datamodel-types';
import EditDecision from '@/definitions/EditDecision';
import clone from '@/store/clone';
import { ValidApplicationStatus } from '@/definitions/ApplicationStatus';
import Term from '@/datamodel/Term';
import ApplicationError from '@/definitions/ApplicationError';
import { BaseState } from './BaseState';
import { Mutations } from 'vuex-smart-module';
import Application from '@/store/Application';

export class RootMutations extends Mutations<Application> {

	public setPropertyPointer( targetProperty: string ): void {
		this.state.targetProperty = targetProperty;
	}

	public setEditFlow( editFlow: string ): void {
		this.state.editFlow = editFlow;
	}

	public setApplicationStatus( status: ValidApplicationStatus ): void {
		this.state.applicationStatus = status;
	}

	public setTargetLabel( label: Term ): void {
		this.state.targetLabel = label;
	}

	public setRenderedTargetReferences( referencesHtml: readonly string[] ): void {
		this.state.renderedTargetReferences = referencesHtml;
	}

	public addApplicationErrors( errors: readonly ApplicationError[] ): void {
		this.state.applicationErrors.push( ...errors );
	}

	public clearApplicationErrors(): void {
		this.state.applicationErrors = [];
	}

	public setEditDecision( editDecision: EditDecision ): void {
		this.state.editDecision = editDecision;
	}

	public setTargetValue( dataValue: DataValue ): void {
		this.state.targetValue = clone( dataValue );
	}

	public setEntityTitle( entityTitle: string ): void {
		this.state.entityTitle = entityTitle;
	}

	public setPageTitle( pageTitle: string ): void {
		this.state.pageTitle = pageTitle;
	}

	public setOriginalHref( orginalHref: string ): void {
		this.state.originalHref = orginalHref;
	}

	public setPageUrl( pageUrl: string ): void {
		this.state.pageUrl = pageUrl;
	}

	public setShowWarningAnonymousEdit( showWarningAnonymousEdit: boolean ): void {
		this.state.showWarningAnonymousEdit = showWarningAnonymousEdit;
	}

	public setAssertUserWhenSaving( assertUserWhenSaving: boolean ): void {
		this.state.assertUserWhenSaving = assertUserWhenSaving;
	}

	public reset(): void {
		// this.state already has all the properties, and Object.assign() invokes setters, so this is reactive
		Object.assign( this.state, new BaseState() );
	}
}
