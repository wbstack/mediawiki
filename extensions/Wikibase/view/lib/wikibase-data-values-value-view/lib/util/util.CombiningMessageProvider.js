this.util = this.util || {};

util.CombiningMessageProvider = ( function() {
	'use strict';

	/**
	 * Providing messages using two `MessageProvider`s
	 *
	 * @class util.CombiningMessageProvider
	 * @since 0.13.0
	 * @license GNU GPL v2+
	 * @author H. Snater < mediawiki@snater.com >
	 * @author Adrian Heine <adrian.heine@wikimedia.de>
	 *
	 * @constructor
	 *
	 * @param {util.MessageProvider} preferredMessageProvider
	 * @param {util.MessageProvider} alternativeMessageProvider
	 */
	function CombiningMessageProvider( preferredMessageProvider, alternativeMessageProvider ) {
		this._a = preferredMessageProvider;
		this._b = alternativeMessageProvider;
	}

	/**
	 * @class util.CombiningMessageProvider
	 */
	CombiningMessageProvider.prototype = {
		constructor: CombiningMessageProvider,

		/**
		 * @property {util.MessageProvider}
		 * @private
		 */
		_a: null,

		/**
		 * @property {util.MessageProvider}
		 * @private
		 */
		_b: null,

		/**
		 * Tries to get a message via the `MessageProvider`s.
		 *
		 * @inheritdoc
		 */
		getMessage: function( key, params ) {
			return this._a.getMessage( key, params ) || this._b.getMessage( key, params );
		}

	};

	return CombiningMessageProvider;

}() );
