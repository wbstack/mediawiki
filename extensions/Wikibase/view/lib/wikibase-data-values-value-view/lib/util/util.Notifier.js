this.util = this.util || {};

util.Notifier = ( function() {
	'use strict';

	/**
	 * Tracks the current notifications for the `Notifier`'s current function.
	 *
	 * @property {string[]}
	 * @ignore
	 */
	var currentNotifications = [];

	/**
	 * Constructor to create an object which takes several callbacks in its constructor. Each
	 * callback is mapped to a keyword. The keyword can be used in a `notify` function which will
	 * then trigger the callback. The notification object itself is immutable. The object will only
	 * hold a reference to the given map though, and won't copy the map. So, if the map changes from
	 * the outside, the notifier will also be affected.
	 * Instantiation also works without using the `new` keyword.
	 *
	 *     @example
	 *     var notifier = util.Notifier( {
	 *         valid: function() { this.current() },
	 *         invalid: function() { this.current() }
	 *     } );
	 *     notifier.notify( 'valid' ); // will alert 'valid'
	 *     notifier.notify( 'invalid' ); // will alert 'invalid'
	 *     notifier.notify( 'whatever' ); // Nothing happens, no notification registered for this one.
	 *
	 * @class util.Notifier
	 * @license GNU GPL v2+
	 * @author Daniel Werner < daniel.a.r.werner@gmail.com >
	 *
	 * @constructor
	 *
	 * @param {Object} [notificationMap={}] Map from notification IDs to callback functions. The
	 *        context of the functions when called by `notify()` is the `Notifier` instance.
	 *
	 * @throws {Error} if notification map is not specified properly.
	 */
	var SELF = function Notifier( notificationMap ) {
		// allow instance without "new":
		if ( !( this instanceof SELF ) ) {
			return new SELF( notificationMap );
		}

		if ( !notificationMap ) {
			notificationMap = {};
		} else if ( typeof notificationMap !== 'object' ) {
			throw new Error( 'Notifier requires a notification map in form of an object' );
		}

		/**
		 * Will trigger a callback related to a given notification string if there is a callback
		 * function defined for that string.
		 *
		 * @param {string} notification
		 * @param {Array} [args=[]] Optional arguments that will be provided to the callback.
		 * @return {boolean} Whether a notification has been sent. false if the notification has not
		 *         been registered in the constructor.
		 */
		this.notify = function( notification, args ) {
			var notifyCallback = notificationMap[ notification ];

			if ( !notifyCallback ) {
				return false;
			}

			// for this.current() to keep track over current notifications
			currentNotifications.push( notification );

			// context of the callback will be the Notifier instance
			notifyCallback.apply( this, args || [] );

			// NOTE: the above might fail with an error. If there is a try outside the notify() call,
			//  the currentNotifications won't be updated after the error got triggered. Putting
			//  the above in a try would be annoying for debugging. currentNotifications having
			//  remnants doesn't have any side-effects as long as we don't use it as an implication
			//  for whether we are inside of a notification right now.

			currentNotifications.pop();
			return true;
		};

		/**
		 * Returns what is currently being notified. Will only return a value when used within a
		 * callback because only within callbacks things are being notified.
		 *
		 * @return {string|null}
		 */
		this.current = function() {
			var current = currentNotifications[ currentNotifications.length - 1 ];
			return current !== undefined ? current : null;
		};
	};

	return SELF;

}() );
