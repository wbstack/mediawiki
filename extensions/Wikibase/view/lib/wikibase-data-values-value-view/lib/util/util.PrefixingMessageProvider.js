this.util = this.util || {};

util.PrefixingMessageProvider = ( function() {
	'use strict';

	/**
	 * Calls another `MessageProvider` with prefixed message keys
	 *
	 * @class util.PrefixingMessageProvider
	 * @since 0.13.0
	 * @license GNU GPL v2+
	 * @author H. Snater < mediawiki@snater.com >
	 * @author Adrian Heine <adrian.heine@wikimedia.de>
	 *
	 * @constructor
	 *
	 * @param {string} prefix
	 * @param {util.MessageProvider} messageProvider
	 */
	function PrefixingMessageProvider( prefix, messageProvider ) {
		this._prefix = prefix;
		this._messageProvider = messageProvider;
	}

	/**
	 * @class util.PrefixingMessageProvider
	 */
	PrefixingMessageProvider.prototype = {
		constructor: PrefixingMessageProvider,

		/**
		 * @property {string}
		 * @private
		 */
		_prefix: null,

		/**
		 * @property {util.MessageProvider}
		 * @private
		 */
		_messageProvider: null,

		/**
		 * Tries to get a message via the provided `MessageProvider`.
		 *
		 * @inheritdoc
		 */
		getMessage: function( key, params ) {
			return this._messageProvider.getMessage( this._prefix + key, params );
		}

	};

	return PrefixingMessageProvider;

}() );
