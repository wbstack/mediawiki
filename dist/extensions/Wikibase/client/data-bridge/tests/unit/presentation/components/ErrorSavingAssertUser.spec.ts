import MessageKeys from '@/definitions/MessageKeys';
import ErrorSavingAssertUser from '@/presentation/components/ErrorSavingAssertUser.vue';
import {
	config,
	shallowMount,
} from '@vue/test-utils';
import { createTestStore } from '../../../util/store';
import { BridgeConfig } from '@/store/Application';
import { ComponentOptions, nextTick } from 'vue';

beforeAll( () => {
	config.global.renderStubDefaultSlot = true;
} );

afterAll( () => {
	config.global.renderStubDefaultSlot = false;
} );

describe( 'ErrorSavingAssertUser', () => {
	const stopAssertingUserWhenSaving = jest.fn();
	const retrySave = jest.fn();
	const goBackFromErrorToReady = jest.fn();
	const store = createTestStore( {
		actions: {
			stopAssertingUserWhenSaving,
			retrySave,
			goBackFromErrorToReady,
		},
		state: {
			config: { usePublish: false } as BridgeConfig,
		},
	} );

	it( 'matches the snapshot', () => {
		const $messages = {
			KEYS: MessageKeys,
			getText: jest.fn().mockReturnValue( "Some 'text" ),
		};

		const wrapper = shallowMount( ErrorSavingAssertUser, {
			propsData: {
				loginUrl: 'https://data-bridge.test/Login',
			},
			global: {
				mocks: { $messages },
				plugins: [ store ],
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'dispatches stopAssertingUserWhenSaving and retrySave when publish without logging in is clicked', async () => {
		const $messages = {
			KEYS: MessageKeys,
			getText: jest.fn().mockReturnValue( "Some 'text" ),
		};

		const wrapper = shallowMount( ErrorSavingAssertUser, {
			propsData: {
				loginUrl: 'https://data-bridge.test/Login',
			},
			global: {
				mocks: { $messages },
				plugins: [ store ],
			},
		} );

		const button = wrapper.findComponent<ComponentOptions>( '.wb-db-error-saving-assertuser__proceed' );
		await button.vm.$emit( 'click' );
		await nextTick();

		expect( stopAssertingUserWhenSaving ).toHaveBeenCalledTimes( 1 );
		expect( retrySave ).toHaveBeenCalledTimes( 1 );
	} );

	// from our point of view, login/app behave identically; difference is tested in browser tests
	it.each( [
		[ 'login' ],
		[ 'back' ],
	] )( 'goes back if the %s button is clicked', async ( buttonName: string ) => {
		const $messages = {
			KEYS: MessageKeys,
			getText: jest.fn().mockReturnValue( "Some 'text" ),
		};

		const wrapper = shallowMount( ErrorSavingAssertUser, {
			propsData: {
				loginUrl: 'https://data-bridge.test/Login',
			},
			global: {
				mocks: { $messages },
				plugins: [ store ],
			},
		} );

		const button = wrapper.findComponent<ComponentOptions>( `.wb-db-error-saving-assertuser__${buttonName}` );
		button.vm.$emit( 'click' );

		expect( goBackFromErrorToReady ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'renders the proceed button using the SAVING_ERROR_ASSERTUSER_SAVE message', () => {
		const saveMessage = 'Save without logging in';
		const messageGet = jest.fn(
			( key: string ) => {
				if ( key === MessageKeys.SAVING_ERROR_ASSERTUSER_SAVE ) {
					return saveMessage;
				}

				return '';
			},
		);

		const wrapper = shallowMount( ErrorSavingAssertUser, {
			propsData: {
				loginUrl: 'https://data-bridge.test/Login',
			},
			global: {
				mocks: {
					$bridgeConfig: { usePublish: false },
					$messages: {
						KEYS: MessageKeys,
						getText: messageGet,
					},
				},
				plugins: [ store ],
			},
		} );

		const button = wrapper.findComponent<ComponentOptions>( '.wb-db-error-saving-assertuser__proceed' );

		expect( messageGet ).toHaveBeenCalledWith( MessageKeys.SAVING_ERROR_ASSERTUSER_SAVE );
		expect( button.props( 'message' ) ).toBe( saveMessage );
	} );

	it( 'renders the proceed button using the SAVING_ERROR_ASSERTUSER_PUBLISH message', () => {
		const publishMessage = 'Publish without logging in';
		const messageGet = jest.fn(
			( key: string ) => {
				if ( key === MessageKeys.SAVING_ERROR_ASSERTUSER_PUBLISH ) {
					return publishMessage;
				}

				return '';
			},
		);
		const localStore = createTestStore( {
			actions: {
				stopAssertingUserWhenSaving,
				retrySave,
				goBackFromErrorToReady,
			},
			state: {
				config: { usePublish: true } as BridgeConfig,
			},
		} );

		const wrapper = shallowMount( ErrorSavingAssertUser, {
			propsData: {
				loginUrl: 'https://data-bridge.test/Login',
			},
			global: {
				mocks: {
					$messages: {
						KEYS: MessageKeys,
						getText: messageGet,
					},
				},
				plugins: [ localStore ],
			},
		} );

		const button = wrapper.findComponent<ComponentOptions>( '.wb-db-error-saving-assertuser__proceed' );

		expect( messageGet ).toHaveBeenCalledWith( MessageKeys.SAVING_ERROR_ASSERTUSER_PUBLISH );
		expect( button.props( 'message' ) ).toBe( publishMessage );
	} );

} );
