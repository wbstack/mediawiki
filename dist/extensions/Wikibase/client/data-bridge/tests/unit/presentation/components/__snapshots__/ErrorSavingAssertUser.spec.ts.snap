// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSavingAssertUser matches the snapshot 1`] = `
<div
  class="wb-db-error-saving-assertuser"
>
  <h2
    class="wb-db-error-saving-assertuser__heading"
  >
    Some 'text
  </h2>
  <icon-message-box-stub
    class="wb-db-error-saving-assertuser__message"
    inline="false"
    type="warning"
  >
    Some 'text
  </icon-message-box-stub>
  <div
    class="wb-db-error-saving-assertuser__buttons"
  >
    <event-emitting-button-stub
      class="wb-db-error-saving-assertuser__proceed"
      disabled="false"
      message="Some 'text"
      newtab="false"
      preventdefault="true"
      size="M"
      squary="false"
      type="primaryProgressive"
    />
    <event-emitting-button-stub
      class="wb-db-error-saving-assertuser__login"
      disabled="false"
      href="https://data-bridge.test/Login"
      message="Some 'text"
      newtab="true"
      preventdefault="false"
      size="M"
      squary="false"
      type="neutral"
    />
    <event-emitting-button-stub
      class="wb-db-error-saving-assertuser__back"
      disabled="false"
      message="Some 'text"
      newtab="false"
      preventdefault="true"
      size="M"
      squary="false"
      type="link"
    />
  </div>
</div>
`;
